### PR TITLE
Support exchange properties

### PR DESCRIPTION
**Pull request by coderzc**
Master Issue: #631

### Motivation

Support exchange properties

### Modifications

Save durable, auto-delete to exchange properties.
Add exchange declare check for the durable and auto-delete parameter.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

Check the box below.

Need to update docs? 

- [ ] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [x] `no-need-doc` 
  
  (Please explain why)
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

**Comment by coderzc** [comment]
@gaoran10 @codelipenghui PTAL

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/impl/PersistentExchange.java [hunk 1]
```diff
@@ -213,20 +215,53 @@ public Topic getTopic(){
...
+        properties.forEach((k, v) -> {
+            switch (k) {
+                case QUEUES:
+                    map.put(k, ExchangeUtil.covertStringValueAsObject(v, List.class));
+                    break;
+                case TYPE:
+                case EXCHANGE:
+                    map.put(k, ExchangeUtil.covertStringValueAsObject(v, String.class));
+                    break;
+                case INDEX:
```
Comment (gaoran10) [inline amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/impl/PersistentExchange.java:235]: Why need this?

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/impl/InMemoryExchange.java [hunk 2]
```diff
@@ -44,11 +44,20 @@ public class InMemoryExchange extends AbstractAmqpExchange {
     private final long currentLedgerId;
     private long currentEntryId;
 
+    private Map<String, Object> properties;
```
Comment (gaoran10) [inline amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/impl/InMemoryExchange.java:47]: Maybe we can add this field in class `AbstractAmqpExchange`.

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/impl/PersistentExchange.java [hunk 3]
```diff
@@ -213,20 +215,53 @@ public Topic getTopic(){
...
+        Map<String, String> properties = this.persistentTopic.getManagedLedger().getProperties();
+        if (properties == null) {
+            return null;
+        }
+        Map<String, Object> map = new HashMap<>();
+
+        properties.forEach((k, v) -> {
+            switch (k) {
+                case QUEUES:
+                    map.put(k, ExchangeUtil.covertStringValueAsObject(v, List.class));
```
Comment (gaoran10) [inline amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/impl/PersistentExchange.java:229]: The `QUEUES` is not a custom property.

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/impl/PersistentExchange.java [hunk 4]
```diff
@@ -213,20 +215,53 @@ public Topic getTopic(){
...
+                    break;
+                case TYPE:
+                case EXCHANGE:
```
Comment (gaoran10) [inline amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/impl/PersistentExchange.java:232]: Same with `QUEUES`.

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeContainer.java [hunk 5]
```diff
@@ -83,7 +95,14 @@ public CompletableFuture<AmqpExchange> asyncGetExchange(NamespaceName namespaceN
             return existingAmqpExchangeFuture;
         } else {
             String topicName = PersistentExchange.getExchangeTopicName(namespaceName, exchangeName);
-            CompletableFuture<Topic> topicCompletableFuture = amqpTopicManager.getTopic(topicName, createIfMissing);
+
+            Map<String, String> newProperties = null;
+            if (createIfMissing && arguments != null) {
```
Comment (gaoran10) [inline amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeContainer.java:100]: Maybe we can serialize declare custom properties to a JSON string, and put it to a specific key.

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeContainer.java [hunk 6]
```diff
@@ -57,7 +68,8 @@ protected ExchangeContainer(AmqpTopicManager amqpTopicManager, PulsarService pul
     public CompletableFuture<AmqpExchange> asyncGetExchange(NamespaceName namespaceName,
                                                             String exchangeName,
                                                             boolean createIfMissing,
-                                                            String exchangeType) {
+                                                            String exchangeType,
+                                                            FieldTable arguments) {
```
Comment (gaoran10) [inline amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeContainer.java:72]: Could we change this to `Map<String, Object>`?

**Review by gaoran10 — COMMENTED** [review]
Great work! Left some comments.

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/impl/InMemoryExchange.java (hunk 2 above)
Comment (coderzc) [reply amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/impl/InMemoryExchange.java:47]: `PersistentExchange` don't need this field.

**Review by gaoran10 — APPROVED** [review]

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/utils/ExchangeUtil.java [hunk 7]
```diff
@@ -53,4 +75,43 @@ public static boolean isDefaultExchange(final String exchangeName) {
...
+        }
+        if (StringUtils.isEmpty(exchangeType)) {
+            throw new ExchangeParameterException("Miss parameter exchange type.");
+        }
+        Map<String, String> props = new HashMap<>();
+        props.put(EXCHANGE, exchangeName);
+        props.put(TYPE, exchangeType);
+        props.put(DURABLE, "" + durable);
+        props.put(AUTO_DELETE, "" + autoDelete);
+        props.put(INTERNAL, "" + internal);
```
Comment (coderzc) [inline amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/utils/ExchangeUtil.java:107]:
```suggestion
        props.put(DURABLE, Double.toString(durable));
        props.put(AUTO_DELETE, Double.toString(autoDelete));
        props.put(INTERNAL, Double.toString(internal));
```

Comment (gaoran10) [reply amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/utils/ExchangeUtil.java:107]: Sorry, I'll resolve this in another PR.

**Pull request merged**